File: Java/Tree/BST to max heap.java
  import java.util.ArrayList;

// problem Link: https://practice.geeksforgeeks.org/problems/bst-to-max-heap/1


/****
 * Approach: first we do inorder traversal(Left,root,right) of given BST, such that we get
 * sorted list of integers.
 * 
 * Then we are doing preorder traversal(Left,right,root) of tree and put that
 * list values inplace of tree values. 
 */

/*****
 * Time Complexity: O(N)
 * Space Complexity: O(N)
 */

class Solution
{
    static ArrayList<Integer> al=new ArrayList<>();
    static int i=0;
    public static void inorder(Node root){
        if(root == null) return ;
        inorder(root.left);
        al.add(root.data);
        inorder(root.right);
    }

    public static void preorder(Node root){
        if(root == null )return; 
        preorder(root.left);
        preorder(root.right);
        root.data= al.get(i++);
    }
    public static void convertToMaxHeapUtil(Node root)
    {
        //code here
        inorder(root);
        preorder(root);
    }
}

File: Python/Tree/BST to max heap.py
#Problem link:https://practice.geeksforgeeks.org/problems/bst-to-max-heap/1

#Time Complexity: O(N)
#Space Complexity: O(N)
class Solution:
        
    def convertToMaxHeapUtil(self, root):
        arr = []
        i = 0
        
        #inOrder traversal
        #to get the values sorted
        def ino(node):
            if not node:
                return
            
            ino(node.left)
            arr.append(node.data)
            ino(node.right)
            
        ino(root)#call the inorder traversal
        
        #preOrder traversal
        #to put the list values
        #inplace of tree values
        def dfs(node):
            nonlocal i
            if not node:
                return
            
            dfs(node.left)
            dfs(node.right)
            node.data=arr[i]
            i+=1
            
        dfs(root)#call the preorder traversal
